### PR TITLE
jetbrains.rider: fix startup

### LINKED ISSUE #190835: JetBrains Rider fails to startup, missing libstdc++.so.6
### Describe the bug

For the last few weeks, Rider hasn't started. I get an dialog with the following text:

```
Internal error. Please refer to https://jb.gg/ide/critical-startup-errors

com.intellij.ide.plugins.StartupAbortedException: Cannot start app
    at com.intellij.idea.StartupUtil.lambda$start$15(StartupUtil.java:265)
    at java.base/java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:986)
    at java.base/java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:970)
    at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
    at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1705)
    at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1692)
    at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
    at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
    at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
    at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
    at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: com.intellij.diagnostic.PluginException: Fatal error initializing 'com.jetbrains.rider.protocol.ProtocolManagerInitializer' [Plugin: com.intellij]
    at com.intellij.serviceContainer.ComponentManagerImpl.handleInitComponentError$intellij_platform_serviceContainer(ComponentManagerImpl.kt:525)
    at com.intellij.serviceContainer.MyComponentAdapter.doCreateInstance(MyComponentAdapter.kt:60)
    at com.intellij.serviceContainer.BaseComponentAdapter.getInstanceUncached(BaseComponentAdapter.kt:119)
    at com.intellij.serviceContainer.BaseComponentAdapter.getInstance(BaseComponentAdapter.kt:73)
    at com.intellij.serviceContainer.BaseComponentAdapter.getInstance$default(BaseComponentAdapter.kt:66)
    at com.intellij.serviceContainer.ComponentManagerImpl$createInitOldComponentsTask$1.run(ComponentManagerImpl.kt:393)
    at com.intellij.idea.ApplicationLoader$prepareStart$1.run(ApplicationLoader.kt:217)
    at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1736)
    at com.intellij.openapi.application.TransactionGuardImpl.runWithWritingAllowed(TransactionGuardImpl.java:209)
    at com.intellij.openapi.application.TransactionGuardImpl.access$100(TransactionGuardImpl.java:21)
    at com.intellij.openapi.application.TransactionGuardImpl$1.run(TransactionGuardImpl.java:191)
    at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:881)
    at com.intellij.openapi.application.impl.ApplicationImpl$3.run(ApplicationImpl.java:513)
    at com.intellij.openapi.application.impl.FlushQueue.doRun(FlushQueue.java:75)
    at com.intellij.openapi.application.impl.FlushQueue.runNextEvent(FlushQueue.java:118)
    at com.intellij.openapi.application.impl.FlushQueue.flushNow(FlushQueue.java:42)
    at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
    at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:776)
    at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
    at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
    at java.base/java.security.AccessController.doPrivileged(Native Method)
    at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
    at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:746)
    at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:402)
    at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
    at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
    at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
    at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
    at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
    at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Caused by: java.lang.IllegalStateException: No protocolHost for the application
    at com.jetbrains.rider.protocol.ApplicationProtocolKt.Y(ApplicationProtocol.kt:41)
    at com.jetbrains.rider.protocol.ApplicationProtocolKt.getProtocolManager(ApplicationProtocol.kt:15)
    at com.jetbrains.rider.protocol.ProtocolManagerInitializer.<init>(ProtocolManager.kt:68)
    at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
    at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
    at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
    at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
    at com.intellij.serviceContainer.ConstructorInjectionKt.instantiateUsingPicoContainer(constructorInjection.kt:47)
    at com.intellij.serviceContainer.ComponentManagerImpl.instantiateClassWithConstructorInjection(ComponentManagerImpl.kt:890)
    at com.intellij.serviceContainer.MyComponentAdapter.doCreateInstance(MyComponentAdapter.kt:35)
    ... 28 more

-----
Your JRE: 11.0.13+0-adhoc..source amd64 (Oracle Corporation)
/nix/store/pmlqxwxcxnkg74fyy4xf9qpg8dld4g99-jetbrains-jdk-11_0_13-b1751.25/lib/openjdk
```

In the console, there is a link to recovering from IDE errors. I've followed those including blowing away my config, log, and cache directories (which lets me go through the initial setup but still fails).

### Steps To Reproduce

1. Include Rider and build system.
2. In the CLI, navigate to a C# project and enter `rider Solution.sln`

### Expected behavior

It starts up after a while and it is usable.

### Screenshots

I pasted above the text of the only dialog box.

### Additional context

Looking through the error log, I saw this entry:

```
java.io.IOException: Cannot run program "/nix/store/n3xb1vr2vyahf628z7y576i0dmf9k6na-rider-2022.2.1/rider/lib/ReSharperHost/linux-x64/Rider.Backend": error=2, No such file or directory
```

This file exists but running `lld` on it show this:

```console
> ldd /nix/store/n3xb1vr2vyahf628z7y576i0dmf9k6na-rider-2022.2.1/rider/lib/ReSharperHost/linux-x64/Rider.Backend
	linux-vdso.so.1 (0x00007fffa19b4000)
	libpthread.so.0 => /nix/store/6w8g7njm4mck5dmjxws0z1xnrxvl81xa-glibc-2.34-115/lib/libpthread.so.0 (0x00007f1bc1286000)
	libdl.so.2 => /nix/store/6w8g7njm4mck5dmjxws0z1xnrxvl81xa-glibc-2.34-115/lib/libdl.so.2 (0x00007f1bc1281000)
	libstdc++.so.6 => not found
	libm.so.6 => /nix/store/6w8g7njm4mck5dmjxws0z1xnrxvl81xa-glibc-2.34-115/lib/libm.so.6 (0x00007f1bc11a8000)
	libgcc_s.so.1 => /nix/store/6w8g7njm4mck5dmjxws0z1xnrxvl81xa-glibc-2.34-115/lib/libgcc_s.so.1 (0x00007f1bc118e000)
	libc.so.6 => /nix/store/6w8g7njm4mck5dmjxws0z1xnrxvl81xa-glibc-2.34-115/lib/libc.so.6 (0x00007f1bc0f8d000)
	/lib64/ld-linux-x86-64.so.2 => /nix/store/6w8g7njm4mck5dmjxws0z1xnrxvl81xa-glibc-2.34-115/lib64/ld-linux-x86-64.so.2 (0x00007f1bc12b1000)
```

The problem seems to be the missing `libstdc++.so.6`.

### Notify maintainers

Rider doesn't appear to have maintainers, but @edwtjo is on a number of packages in that file.

### Metadata

```console
> nix-shell -p nix-info --run "nix-info -m"
this path will be fetched (0.00 MiB download, 0.00 MiB unpacked):
  /nix/store/q124axggd6wl6fh9508sxgkknkifrl0z-nix-info
copying path '/nix/store/q124axggd6wl6fh9508sxgkknkifrl0z-nix-info' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.67, NixOS, 22.11 (Raccoon)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.11.0`
 - channels(root): `"nixos-22.05pre358459.3e072546ea9"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by raphaelr**
Rider 2022.2 changed the way it invokes Rider.Backend, this adds the
required patching so that startup succeeds again.

Fixes: #190835

###### Description of changes

Currently Rider does not start at all, and shows [this error instead](https://gist.github.com/raphaelr/3caef4463ad6d6951aea8ce9c6b855c3)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by raphaelr** [comment]
Requesting review from @GenericNerdyUsername or @drupol since this package has no current maintainer.

**Comment by GenericNerdyUsername** [comment]
It launches, but thats all I can test as I don't use rider

**Review by GenericNerdyUsername — APPROVED** [review]

**Comment by drupol** [comment]
Sorry I don't use Rider.

File: pkgs/applications/editors/jetbrains/default.nix [hunk 1]
```diff
@@ -222,13 +224,17 @@ let
...
       };
     }).overrideAttrs (attrs: {
       postPatch = lib.optionalString (!stdenv.isDarwin) (attrs.postPatch + ''
+        interp="$(cat $NIX_CC/nix-support/dynamic-linker)"
+        patchelf --set-interpreter $interp lib/ReSharperHost/linux-x64/Rider.Backend
+
         rm -rf lib/ReSharperHost/linux-x64/dotnet
         mkdir -p lib/ReSharperHost/linux-x64/dotnet/
         ln -s ${dotnet-sdk_6}/bin/dotnet lib/ReSharperHost/linux-x64/dotnet/dotnet
+        ln -s ${dotnet-sdk_6}/{host,shared} lib/ReSharperHost/linux-x64/dotnet/
```
Comment (huantianad) [inline pkgs/applications/editors/jetbrains/default.nix:237]: Should we just link the entirety of `dotnet-sdk_6`, i.e. `ln -s ${pkgs.dotnet-sdk_6} lib/ReSharperHost/linux-x64/dotnet`?

**Comment by huantianad** [comment]
Seems to work in general with some light testing, I am getting java version errors though that should be fixed with #183763.

**Review by huantianad — APPROVED** [review]

**Review by khrj — APPROVED** [review]

**Pull request merged**